src/middlewared/middlewared/service.py

```diff
@@ -556,6 +562,93 @@ async def check_dependencies(self, id, ignored=None):
                             {'dependencies': list(dependencies.values())})
 
 
+class SharingTaskService(CRUDService):
+
+    path_field = 'path'
+    enabled_field = 'enabled'
+    locked_field = 'locked'
+    service_type = NotImplemented
+    locked_alert_class = NotImplemented
+
+    @private
+    async def sharing_task_extend_context(self, extra):
+        return {
+            'locked_datasets': await self.middleware.call('zfs.dataset.locked_datasets'),
+            'service_extend': (await self.middleware.call(self._config.datastore_extend_context, extra))
+            if self._config.datastore_extend_context else {}
+        }
+
+    @private
+    async def validate_path_field(self, data, schema, verrors=None):
+        verrors = verrors or ValidationErrors()
+        await check_path_resides_within_volume(
+            verrors, self.middleware, f'{schema}.{self.path_field}', data.get(self.path_field)
+        )
+        if data[self.enabled_field] and await self.middleware.call(
+            'pool.dataset.path_in_locked_datasets', data[self.path_field]
+        ):
+            verrors.add(
+                f'{schema}.{self.path_field}',
+                'Underlying resource is locked.'
+            )
+            verrors.add(
+                f'{schema}.{self.enabled_field}',
+                f'{self.service_type.capitalize()} must be disabled to configure the specified resource(s).'
+            )
+        return verrors
+
+    @private
+    async def sharing_task_determine_locked(self, data, locked_datasets):
+        return await self.middleware.call(
+            'pool.dataset.path_in_locked_datasets', data[self.path_field], locked_datasets
+        )
+
+    @private
+    async def sharing_task_extend(self, data, context):
+        args = [data] + ([context['service_extend']] if self._config.datastore_extend_context else [])
+
+        if self._config.datastore_extend:
+            data = await self.middleware.call(self._config.datastore_extend, *args)
+
+        data[self.locked_field] = await self.middleware.call(
+            f'{self._config.namespace}.sharing_task_determine_locked', data, context['locked_datasets']
+        )
+
+        return data
+
+    async def get_options(self, options):
+        return {
+            **(await super().get_options(options)),
+            'extend': f'{self._config.namespace}.sharing_task_extend',
+            'extend_context': f'{self._config.namespace}.sharing_task_extend_context',
+        }
+
+    @private
+    async def remove_locked_alert(self, share_task_id):
+        await self.middleware.call('alert.oneshot_delete', self.locked_alert_class, share_task_id)
+
+    @pass_app(rest=True)
+    async def update(self, app, id, data):
+        rv = await super().update(app, id, data)
+        if not rv[self.enabled_field]:
```

Comment:
Will `alert.oneshot_delete` be called if we change the path from locked dataset to unlocked/unencrypted dataset?